### PR TITLE
[clang-format] Reorder TokenAnnotator::canBreakBefore

### PR DESCRIPTION
**Pull request by gedare**
Move the checks related to breaking before right braces and right parens earlier to avoid conflicting checks that prevent breaking based on the left-hand token.  This allows properly formatting declarations with pointers and references at a minimum.

### CODE REVIEW

**Comment by gedare** [comment]
Without this patch, the test case on `main`:
```
[ RUN      ] FormatTest.AlignsAfterOpenBracket
/home/gedare/rtems/llvm-project/clang/unittests/Format/FormatTestBase.h:90: Failure
Expected equality of these values:
  ExpectedCode
    Which is: "void foo(\n    void (*foobarpntr)(\n        aaaaaaaaaaaaaaaaaa *,\n        bbbbbbbbbbbbbb *,\n        cccccccccccccccccccc *,\n        dddddddddddddddddd *\n    )\n);"
  FormattedCode
    Which is: "void foo(\n    void (*foobarpntr)(aaaaaaaaaaaaaaaaaa *, bbbbbbbbbbbbbb *, cccccccccccccccccccc *, dddddddddddddddddd *)\n);"
With diff:
@@ -1,8 +1,3 @@
 void foo(
-    void (*foobarpntr)(
-        aaaaaaaaaaaaaaaaaa *,
-        bbbbbbbbbbbbbb *,
-        cccccccccccccccccccc *,
-        dddddddddddddddddd *
-    )
+    void (*foobarpntr)(aaaaaaaaaaaaaaaaaa *, bbbbbbbbbbbbbb *, cccccccccccccccccccc *, dddddddddddddddddd *)
 );

Google Test trace:
/home/gedare/rtems/llvm-project/clang/unittests/Format/FormatTest.cpp:9451: void foo(
    void (*foobarpntr)(
        aaaaaaaaaaaaaaaaaa *,
        bbbbbbbbbbbbbb *,
        cccccccccccccccccccc *,
        dddddddddddddddddd *
    )
);

[  FAILED  ] FormatTest.AlignsAfterOpenBracket (379 ms)
```

File: clang/lib/Format/TokenAnnotator.cpp [hunk 1]
```diff
@@ -6105,6 +6105,33 @@ bool TokenAnnotator::canBreakBefore(const AnnotatedLine &Line,
       return false;
   }
 
+  // We only break before r_brace if there was a corresponding break before
+  // the l_brace, which is tracked by BreakBeforeClosingBrace.
```
Comment (owenca) [inline clang/lib/Format/TokenAnnotator.cpp:6109]:
```suggestion
  // We only break before an r_brace if there was a corresponding break after
  // the matching l_brace, which is tracked by BreakBeforeClosingBrace.
```

File: clang/lib/Format/TokenAnnotator.cpp [hunk 2]
```diff
@@ -6105,6 +6105,33 @@ bool TokenAnnotator::canBreakBefore(const AnnotatedLine &Line,
...
+    const FormatToken *Previous = Right.MatchingParen->Previous;
+    return !(Previous && (Previous->is(tok::kw_for) || Previous->isIf()));
+  }
+
+  if (Left.isOneOf(tok::r_paren, TT_TrailingAnnotation) &&
+      Right.is(TT_TrailingAnnotation) &&
+      Style.AlignAfterOpenBracket == FormatStyle::BAS_BlockIndent) {
+    return false;
+  }
   if (Left.is(tok::at))
```
Comment (owenca) [inline clang/lib/Format/TokenAnnotator.cpp:6135]:
```suggestion

  if (Left.is(tok::at))
```

File: clang/unittests/Format/FormatTest.cpp [hunk 3]
```diff
@@ -9383,6 +9383,13 @@ TEST_F(FormatTest, AlignsAfterOpenBracket) {
       "        aaaaaaaaaaaaaaaaaaaa(aaaaaaaaaaaaaaaaa, aaaaaaaaaaaaaaaa)) &&\n"
       "    aaaaaaaaaaaaaaaa);",
       Style);
+  verifyFormat("void foo(\n"
+               "    void (*foobarpntr)(\n"
+               "        aaaaaaaaaaaaaaaaaa *,\n"
+               "        bbbbbbbbbbbbbb *,\n"
+               "        cccccccccccccccccccc *,\n"
+               "        dddddddddddddddddd *));",
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:9391]: We probably can do without this test, which would pass without this patch.

File: clang/lib/Format/TokenAnnotator.cpp (hunk 1 above)
Comment (gedare) [reply clang/lib/Format/TokenAnnotator.cpp]: Actually, this comment is outdated. I've pushed your suggestion with a little bit more correction.

File: clang/lib/Format/TokenAnnotator.cpp [hunk 4]
```diff
@@ -6105,6 +6105,35 @@ bool TokenAnnotator::canBreakBefore(const AnnotatedLine &Line,
       return false;
   }
 
+  // We can break before an r_brace if there was a corresponding break after
+  // the l_brace, which is tracked by BreakBeforeClosingBrace, or if we are
+  // in a block indented initialization list.
```
Comment (owenca) [inline clang/lib/Format/TokenAnnotator.cpp:6110]:
```suggestion
  // We can break before an r_brace if there was a break after the matching
  // l_brace, which is tracked by BreakBeforeClosingBrace, or if we are in a
  // block-indented initialization list.
```

**Review by owenca — APPROVED** [review]

**Pull request merged**